packages/x-charts/src/internals/useAnimate.ts

```diff
@@ -0,0 +1,123 @@
+'use client';
+import useId from '@mui/utils/useId';
+import * as React from 'react';
+import { interrupt, Transition } from '@mui/x-charts-vendor/d3-transition';
+import { select } from '@mui/x-charts-vendor/d3-selection';
+import { shallowEqual } from './shallowEqual';
+import { ANIMATION_DURATION_MS, ANIMATION_TIMING_FUNCTION_JS } from '../constants';
+import { useIsomorphicLayoutEffect } from './useIsomorphicLayoutEffect';
+
+export function useAnimate<Props extends {}, Elem extends Element>(
+  props: Props,
+  {
+    createInterpolator,
+    applyProps,
+    skip,
+    initialProps,
+  }: {
+    createInterpolator: (lastProps: Props, newProps: Props) => (t: number) => Props;
+    applyProps: (element: Elem, props: Props) => void;
+    skip?: boolean;
+    initialProps?: Props;
+  },
+) {
+  const transitionName = `MuiUseAnimate-${useId()}`;
+  /* A transition is interrupted when:
+   * T1. The transitioned element changes;
+   * T2. The transitioned element is unmounted;
+   * T3. The component calling this hook is unmounted. */
+  const lastInterpolatedPropsRef = React.useRef(initialProps ?? props);
+  const transitionRef = React.useRef<Transition<Elem, unknown, null, undefined>>(null);
+  const elementRef = React.useRef<Elem>(null);
+  const elementUnmountedRef = React.useRef(false);
+  const lastPropsRef = React.useRef(props);
+
+  useIsomorphicLayoutEffect(() => {
+    return () => {
+      /* T3. Stop the transition if the component that calls this hook is unmounted. */
+      const lastElement = elementRef.current;
+
+      if (lastElement) {
+        interrupt(lastElement, transitionName);
```

Comment:
This breaks in Strict Mode. D3 transitions aren't resumable, so if we interrupt them we need to start another one, which we aren't at the moment.

Reply:
Due to this, I think I'll need to implement resumable transitions, which D3 transitions doesn't support. 

The idea is to store a transition in a ref and stop it on ref cleanups. When the ref callback is called again with an element, we check if it's the same DOM element and same props: if so, resume transition; if not, start new transition.